### PR TITLE
Sasha/56 crud http wrappers

### PR DESCRIPTION
**Pull request by alex-a-renoire**
issue #56

### CODE REVIEW

**Review by DrIhor — APPROVED** [review]

File: pkg/userGRPC/controller/controller.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+		return fmt.Errorf("%s: %w", message, customerrors.NotFound)
+	}
+
+	if st.Code() == codes.AlreadyExists {
+		return fmt.Errorf("%s: %w", message, customerrors.AlreadyExists)
+	}
+
+	if st.Code() == codes.DeadlineExceeded {
+		return fmt.Errorf("%s: %w", message, customerrors.DeadlineExceeded)
+	}
```
Comment (DrIhor) [inline pkg/userGRPC/controller/controller.go]: Switch.

File: pkg/http/handler/handler.go [hunk 2]
```diff
@@ -69,28 +86,275 @@ func (h HTTPHandler) reportError(w http.ResponseWriter, status int, err error) {
...
+		return
+	}
+
+	accountID, err := h.UsersService.CreateUser(user.Name)
+	if err != nil {
+		h.reportError(w, http.StatusInternalServerError, err)
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/json")
```
Comment (DrIhor) [inline pkg/http/handler/handler.go]: Can add some headers to middleware to not use each time.

File: pkg/accountGRPC/controller/controller.go [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+		return fmt.Errorf("%s: %w", message, customerrors.NotFound)
+	}
+
+	if st.Code() == codes.AlreadyExists {
+		return fmt.Errorf("%s: %w", message, customerrors.AlreadyExists)
+	}
+
+	if st.Code() == codes.DeadlineExceeded {
+		return fmt.Errorf("%s: %w", message, customerrors.DeadlineExceeded)
+	}
```
Comment (DrIhor) [inline pkg/accountGRPC/controller/controller.go]: Switch too.

Comment (DrIhor) [reply pkg/accountGRPC/controller/controller.go]: Or create new service to work with errors.

File: pkg/http/handler/handler.go [hunk 4]
```diff
@@ -69,28 +86,275 @@ func (h HTTPHandler) reportError(w http.ResponseWriter, status int, err error) {
...
+	router.HandleFunc("/users/{id}", h.GetUser).Methods("GET")
+	router.HandleFunc("/users", h.ListUsers).Methods("GET")
+	router.HandleFunc("/users/{id}", h.UpdateUser).Methods("PUT")
+	router.HandleFunc("/users/{id}", h.DeleteUser).Methods("DELETE")
+
+	router.HandleFunc("/accounts", h.AddAccount).Methods("POST")
+	router.HandleFunc("/accounts/{id}", h.GetAccount).Methods("GET")
+	router.HandleFunc("/accounts/{id}", h.UpdateAccount).Methods("PUT")
+	router.HandleFunc("/accounts/{id}", h.DeleteAccount).Methods("DELETE")
+	router.HandleFunc("/accounts", h.ListAccounts).Methods("GET")
```
Comment (DrIhor) [inline pkg/http/handler/handler.go:107]: Can split this part with sub routers if too many endpoints.

Comment (DrIhor) [reply pkg/http/handler/handler.go:107]: Or create 2 functions with endpoints(`GetAccountRouter` and `GetUserRouter`)

**Comment by DrIhor** [comment]
Check other [pr](https://github.com/stasBigunenko/monorepa/pull/71). Same code:(

File: pkg/http/handler/handler.go (hunk 4 above)
Comment (romanilchyshyn) [reply pkg/http/handler/handler.go:107]: Have no idea why is it too many. IMO looks OK.

**Comment by alex-a-renoire** [comment]
> Check other [pr](https://github.com/stasBigunenko/monorepa/pull/71). Same code:(

Didn't understand

**Comment by DrIhor** [comment]
> > Check other [pr](https://github.com/stasBigunenko/monorepa/pull/71). Same code:(
> 
> Didn't understand

Approximately the same code. Therefore look as same things in 2 pr.

File: customErrors/httperrors.go [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+	Message: "failed to parse uuid",
+}
+
+var RWError = HTTPError{
+	Message: "failed to read / write to io",
+}
+
+var JSONError = HTTPError{
+	Message: "failed to marshal / unmarshal json",
+}
```
Comment (DrIhor) [inline customErrors/httperrors.go:21]: Can connect vars(examlple: `customErrors/grpcerrors.go`)

File: pkg/accountGRPC/controller/controller.go [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+	for _, account := range resp.Accounts {
+		accountID, err := uuid.Parse(account.Id)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse account ID: %s, %w", err.Error(), customerrors.ParseError)
+		}
+
+		userID, err := uuid.Parse(account.UserID)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse account ID: %s, %w", err.Error(), customerrors.ParseError)
+		}
```
Comment (DrIhor) [inline pkg/accountGRPC/controller/controller.go:107]: You have user ID as enter parameter. Why parse each time???

File: pkg/auth/middleware/headers.go [hunk 7]
```diff
@@ -4,9 +4,22 @@ import "net/http"
...
+		w.Header().Set("Access-Control-Allow-Methods", "GET, POST, OPTIONS,PUT")
+		w.Header().Set("Access-Control-Allow-Headers", "Origin, Content-Type")
+
+		if r.Method == "OPTIONS" {
+			return
+		}
+
+		next.ServeHTTP(w, r)
+	})
+}
```
Comment (DrIhor) [inline pkg/auth/middleware/headers.go:25]: From other pr...(was mergerd)

**Comment by DrIhor** [comment]
Many other pr data(((

**Pull request closed without merging**